### PR TITLE
Impl mobilenetv2

### PR DESCRIPTION
**Pull request by okdshin**
This PR contains MobileNetV2 implementation and trained parameter converter.
Trained parameters are from TensorFlow's implementation.
https://github.com/tensorflow/models/tree/master/research/slim/nets/mobilenet

### CODE REVIEW

File: examples/mobilenet/README.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# MobileNet
+
+For evaluation, please go to [`examples/classification`](https://github.com/chainer/chainercv/tree/master/examples/classification).
+
+## Convert Caffe model
```
Comment (knorth55) [inline examples/mobilenet/README.md]: tensorflow

File: examples/mobilenet/tfckpt2npz.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+            'MobilenetV2/Logits/Conv2d_1c_1x1/weights'),
+        bias=ckpt_reader.get_tensor('MobilenetV2/Logits/Conv2d_1c_1x1/biases'))
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        'model_name', choices=('mobilenetv2', ), default='mobilenetv2')
+    parser.add_argument('pretrained_model')
+    parser.add_argument('--n_class', type=int, default=1001)
```
Comment (knorth55) [inline examples/mobilenet/tfckpt2npz.py]: please use `--n-class`, no underscore in argument

File: chainercv/links/model/mobilenet/tf_convolution_2d.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
+import numpy as np
+import chainer
+from chainer.links import Convolution2D
+from chainer.functions import pad
+from chainer.utils import conv
```
Comment (knorth55) [inline chainercv/links/model/mobilenet/tf_convolution_2d.py]: add empty line after `numpy` import
```
import numpy as np

import chainer
from chainer.links import Convolution2D
from chainer.functions import pad
from chainer.utils import conv
```

File: examples/classification/eval_imagenet.py [hunk 4]
```diff
@@ -44,8 +44,12 @@ def main():
...
     n_class = len(label_names)
     iterator = iterators.MultiprocessIterator(
-        dataset, args.batchsize, repeat=False, shuffle=False,
-        n_processes=6, shared_mem=300000000)
+        dataset,
+        args.batchsize,
+        repeat=False,
+        shuffle=False,
+        n_processes=6,
+        shared_mem=300000000)
```
Comment (knorth55) [inline examples/classification/eval_imagenet.py]: it looks no modification.
do we really need this change?

File: examples/mobilenet/tfckpt2npz.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
+import argparse
+
+import numpy as np
+
+import chainer
+
+import tensorflow as tf
+
+from chainercv.links import MobileNetV2
```
Comment (knorth55) [inline examples/mobilenet/tfckpt2npz.py]: split imports into three sections

```
import argparse
import numpy as np

import chainer
import tensorflow as tf

from chainercv.links import MobileNetV2
```

File: examples/mobilenet/tfckpt2npz.py (hunk 2 above)
Comment (knorth55) [reply examples/mobilenet/tfckpt2npz.py]: please format other arguments, too

File: examples/mobilenet/README.md [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+Convert TensorFlow's `*.ckpt` to `*.npz`.
+
+```
+$ python caffe2npz.py mobilenetv2 <source>.ckpt <target>.npz
```
Comment (Hakuyume) [inline examples/mobilenet/README.md]: `tf2npz.py`?

Comment (Hakuyume) [reply examples/mobilenet/README.md]: Ah, you are using``tfckpt2npz.py` as the name of conversion script.

Comment (Hakuyume) [reply examples/mobilenet/README.md]: For the consistency with `DeepLab`, how about renaming it `tf2npz.py` ?
https://github.com/chainer/chainercv/blob/master/examples/deeplab/tf2npz.py

File: examples/classification/eval_imagenet.py [hunk 7]
```diff
@@ -28,11 +29,10 @@ def main():
...
     parser.add_argument(
-        '--model', choices=(
-            'vgg16',
-            'resnet50', 'resnet101', 'resnet152',
-            'se-resnet50', 'se-resnet101', 'se-resnet152',
-            'se-resnext50', 'se-resnext101'))
+        '--model',
+        choices=('vgg16', 'resnet50', 'resnet101', 'resnet152', 'se-resnet50',
+                 'se-resnet101', 'se-resnet152', 'se-resnext50',
+                 'se-resnext101', 'mobilenetv2'))
```
Comment (Hakuyume) [inline examples/classification/eval_imagenet.py]: 'mobilenet_v2'

Comment (23pointsNorth) [reply examples/classification/eval_imagenet.py]: Or 'mobilenet-v2'?

File: chainercv/links/model/mobilenet/tf_convolution_2d.py (hunk 3 above)
Comment (okdshin) [reply chainercv/links/model/mobilenet/tf_convolution_2d.py]: OK

File: examples/mobilenet/tfckpt2npz.py (hunk 2 above)
Comment (okdshin) [reply examples/mobilenet/tfckpt2npz.py]: OK

File: examples/mobilenet/README.md (hunk 1 above)
Comment (okdshin) [reply examples/mobilenet/README.md]: OK

File: examples/mobilenet/tfckpt2npz.py (hunk 5 above)
Comment (okdshin) [reply examples/mobilenet/tfckpt2npz.py]: OK

File: examples/mobilenet/README.md (hunk 6 above)
Comment (okdshin) [reply examples/mobilenet/README.md]: DeepLab's tf2npz.py uses tensorflow's frozen model, but this tfckpt2npz uses tf's checkpoint. Frozen model and checkpoint are different. Frozen model is removed some BNs so not to be used for training.

File: examples/classification/eval_imagenet.py (hunk 4 above)
Comment (okdshin) [reply examples/classification/eval_imagenet.py]: I applied yapf. Should I make it back to before?

File: examples/classification/eval_imagenet.py (hunk 7 above)
Comment (okdshin) [reply examples/classification/eval_imagenet.py]: I willl modify it 'mobilenet-v2'

File: examples/classification/eval_imagenet.py (hunk 4 above)
Comment (knorth55) [reply examples/classification/eval_imagenet.py]: yes, please make it back.
it does not exceed any line length limitation

File: examples/classification/eval_imagenet.py (hunk 7 above)
Comment (Hakuyume) [reply examples/classification/eval_imagenet.py]: Both YOLO and DeepLab use underscore to separate its name and version.  So this should be `mobilenet_v2`.

File: examples/mobilenet/README.md [hunk 8]
```diff
@@ -2,12 +2,12 @@
 
 For evaluation, please go to [`examples/classification`](https://github.com/chainer/chainercv/tree/master/examples/classification).
 
-## Convert Caffe model
+## Convert TensorFlow model
 Convert TensorFlow's `*.ckpt` to `*.npz`.
 
 ```
-$ python caffe2npz.py mobilenetv2 <source>.ckpt <target>.npz
+$ pythontfckpt2npz.py mobilenetv2 <source>.ckpt <target>.npz
```
Comment (Hakuyume) [inline examples/mobilenet/README.md]: `python tfckpt2npz.py`

File: chainercv/links/model/mobilenet/__init__.py [hunk 9]
```diff
@@ -0,0 +1,4 @@
+from chainercv.links.model.mobilenet.tf_convolution_2d import TFConvolution2D  # NOQA
+from chainercv.links.model.mobilenet.tf_conv_2d_bn_activ import TFConv2DBNActiv  # NOQA
+from chainercv.links.model.mobilenet.expanded_conv_2d import ExpandedConv2D  # NOQA
+from chainercv.links.model.mobilenet.mobilenetv2 import MobileNetV2  # NOQA
```
Comment (Hakuyume) [inline chainercv/links/model/mobilenet/__init__.py]: `mobilenet_v2`

File: examples/mobilenet/README.md (hunk 6 above)
Comment (Hakuyume) [reply examples/mobilenet/README.md]: OK

**Comment by knorth55** [comment]
travis fails because of `flake8` and test failure.
flake8
```
$ flake8 .
./chainercv/links/__init__.py:28:1: H306  imports not in alphabetical order (chainercv.links.model.yolo.yolov3, chainercv.links.model.mobilenet.mobilenetv2)
./chainercv/links/model/mobilenet/__init__.py:2:1: H306  imports not in alphabetical order (chainercv.links.model.mobilenet.tf_convolution_2d.tfconvolution2d, chainercv.links.model.mobilenet.tf_conv_2d_bn_activ.tfconv2dbnactiv)
./chainercv/links/model/mobilenet/__init__.py:3:1: H306  imports not in alphabetical order (chainercv.links.model.mobilenet.tf_conv_2d_bn_activ.tfconv2dbnactiv, chainercv.links.model.mobilenet.expanded_conv_2d.expandedconv2d)
./chainercv/links/model/mobilenet/expanded_conv_2d.py:3:1: F401 'BatchNormalization' imported but unused
./chainercv/links/model/mobilenet/expanded_conv_2d.py:12:80: E501 line too long (100 > 79 characters)
./chainercv/links/model/mobilenet/expanded_conv_2d.py:17:80: E501 line too long (98 > 79 characters)
./chainercv/links/model/mobilenet/expanded_conv_2d.py:18:80: E501 line too long (93 > 79 characters)
./chainercv/links/model/mobilenet/expanded_conv_2d.py:19:80: E501 line too long (97 > 79 characters)
./chainercv/links/model/mobilenet/expanded_conv_2d.py:20:80: E501 line too long (104 > 79 characters)
./chainercv/links/model/mobilenet/expanded_conv_2d.py:21:80: E501 line too long (100 > 79 characters)
./chainercv/links/model/mobilenet/mobilenetv2.py:7:80: E501 line too long (99 > 79 characters)
./chainercv/links/model/mobilenet/mobilenetv2.py:12:43: H301  one import per line
./chainercv/links/model/mobilenet/mobilenetv2.py:12:80: E501 line too long (80 > 79 characters)
./chainercv/links/model/mobilenet/mobilenetv2.py:13:1: F401 'conv' imported but unused
./chainercv/links/model/mobilenet/mobilenetv2.py:16:1: H306  imports not in alphabetical order (chainercv.links.model.mobilenet.tf_convolution_2d.tfconvolution2d, chainercv.links.model.mobilenet.tf_conv_2d_bn_activ.tfconv2dbnactiv)
./chainercv/links/model/mobilenet/mobilenetv2.py:17:1: H306  imports not in alphabetical order (chainercv.links.model.mobilenet.tf_conv_2d_bn_activ.tfconv2dbnactiv, chainercv.links.model.mobilenet.expanded_conv_2d.expandedconv2d)
./chainercv/links/model/mobilenet/mobilenetv2.py:18:65: H301  one import per line
./chainercv/links/model/mobilenet/mobilenetv2.py:18:80: E501 line too long (88 > 79 characters)
./chainercv/links/model/mobilenet/mobilenetv2.py:73:80: E501 line too long (92 > 79 characters)
./chainercv/links/model/mobilenet/mobilenetv2.py:113:80: E501 line too long (82 > 79 characters)
./chainercv/links/model/mobilenet/mobilenetv2.py:118:25: E262 inline comment should start with '# '
./chainercv/links/model/mobilenet/mobilenetv2.py:118:26: H101  Use TODO(NAME)
./chainercv/links/model/mobilenet/mobilenetv2.py:119:26: H101  Use TODO(NAME)
./chainercv/links/model/mobilenet/mobilenetv2.py:120:26: H101  Use TODO(NAME)
./chainercv/links/model/mobilenet/mobilenetv2.py:204:80: E501 line too long (100 > 79 characters)
./chainercv/links/model/mobilenet/tf_conv_2d_bn_activ.py:58:80: E501 line too long (92 > 79 characters)
./chainercv/links/model/mobilenet/tf_convolution_2d.py:5:1: H306  imports not in alphabetical order (chainer.links.convolution2d, chainer.functions.pad)
./chainercv/links/model/mobilenet/tf_convolution_2d.py:47:80: E501 line too long (89 > 79 characters)
./chainercv/links/model/mobilenet/tf_convolution_2d.py:61:37: E261 at least two spaces before inline comment
./chainercv/links/model/mobilenet/tf_convolution_2d.py:62:80: E501 line too long (105 > 79 characters)
./chainercv/links/model/mobilenet/util.py:13:80: E501 line too long (80 > 79 characters)
./examples/mobilenet/tfckpt2npz.py:2:1: F401 'np' imported but unused
The command "flake8 ." exited with 1.
```

```
args = (<chainer.testing.parameterized.TestFeaturePredictor_param_71 testMethod=test_prepare>,)
kwargs = {}
e = AttributeError("'FeaturePredictor' object has no attribute 'scale'",)
s = <StringIO.StringIO instance at 0x7f8ef10a75a8>, k = 'crop_size'
v = (256, 192)
e_new = AssertionError("Parameterized test failed.\n\nBase test method: TestFeaturePre...192)\n\nAttributeError: 'FeaturePredictor' object has no attribute 'scale'\n",)
    @functools.wraps(method)
    def wrap(*args, **kwargs):
        try:
            return method(*args, **kwargs)
        except unittest.SkipTest:
            raise
        except Exception as e:
            s = six.StringIO()
            s.write('Parameterized test failed.\n\n')
            s.write('Base test method: {}.{}\n'.format(
                base.__name__, method.__name__))
            s.write('Test parameters:\n')
            for k, v in six.iteritems(param):
                s.write('  {}: {}\n'.format(k, v))
            s.write('\n')
            s.write('{}: {}\n'.format(type(e).__name__, e))
            e_new = AssertionError(s.getvalue())
            if sys.version_info < (3,):
>               six.reraise(AssertionError, e_new, sys.exc_info()[2])
```

File: examples/mobilenet/README.md (hunk 8 above)
Comment (okdshin) [reply examples/mobilenet/README.md]: OK

File: examples/classification/eval_imagenet.py (hunk 7 above)
Comment (okdshin) [reply examples/classification/eval_imagenet.py]: I see. I will modify it 'mobilenet_v2'.

File: examples/mobilenet/README.md [hunk 10]
```diff
@@ -0,0 +1,13 @@
+# MobileNet
+
+For evaluation, please go to [`examples/classification`](https://github.com/chainer/chainercv/tree/master/examples/classification).
```
Comment (Hakuyume) [inline examples/mobilenet/README.md]: Please update `examples/classification` to contain the performance of this model. Also, please compare the performance of this implementation and that of the original implementation.

File: examples/classification/eval_imagenet.py (hunk 4 above)
Comment (okdshin) [reply examples/classification/eval_imagenet.py]: OK

File: chainercv/links/model/mobilenet/__init__.py (hunk 9 above)
Comment (okdshin) [reply chainercv/links/model/mobilenet/__init__.py]: OK

File: examples/mobilenet/README.md (hunk 10 above)
Comment (okdshin) [reply examples/mobilenet/README.md]: OK

Comment (okdshin) [reply examples/mobilenet/README.md]: I wrote.

File: chainercv/links/model/mobilenet/expanded_conv_2d.py [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+            Pad of project conv filter application.
+        initialW (callable): Initial weight value used in
+            the convolutional layers.
+        bn_kwargs (dict): Keyword arguments passed to initialize
+            :class:`chainer.links.BatchNormalization`.
+    """
+
+    def __init__(self,
+                 out_channels,
+                 in_channels,
```
[the remark this answers is not in the document]
Comment (okdshin) [reply chainercv/links/model/mobilenet/expanded_conv_2d.py]: OK

**Comment by yuyu2172** [comment]
pfnCI, test this please

**Comment by okdshin** [comment]
Sorry, there are no tests. I will add.

**Comment by okdshin** [comment]
pfnCI, test this please

**Comment by knorth55** [comment]
can you please add more tests of following code?
- `chainercv/links/model/mobilenet/expanded_conv_2d.py`
- `chainercv/links/model/mobilenet/tf_conv_2d_bn_activ.py`
- `chainercv/links/model/mobilenet/tf_convolution_2d.py`
- `chainercv/links/model/feature_predictor.py` with `scale param

**Comment by okdshin** [comment]
I added tests.

File: chainercv/links/model/mobilenet/tf_convolution_2d.py [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+
+    @property
+    def b(self):
+        return self.conv.b
+
+    def forward(self, x):
+        if self.padding is None:
+            return self.conv(x)
+        else:
+            return self.conv(self.padding(x))
```
Comment (yuyu2172) [inline chainercv/links/model/mobilenet/tf_convolution_2d.py]: Can we not use `F.padding` by passing `self.padding` to `pad` option of Convolution?
Using two operations makes implementation slower.

File: chainercv/links/model/mobilenet/mobilenet_v2.py [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+                    'url':
+                    'https://chainercv-models.preferred.jp/mobilenet_v2_depth_multiplier_1.0_imagenet_converted_2019_05_13.npz',  # NOQA
+                }
+            },
+            1.4: {
+                'imagenet': {
+                    'param': {
+                        'n_class': 1001,  # first element is background
+                        'mean': _tf_mobilenetv2_mean,
+                        'scale': _tf_mobilenetv2_scale,
```
Comment (yuyu2172) [inline chainercv/links/model/mobilenet/mobilenet_v2.py]: Can we rescale input value for TF pretraiend model to `[0, 255]` instead of `[0, 1]`?
You can do it by scaling the `W` of the first conv layer when converting pretrained model.
This simplifies the implementation because we no longer need `scale` option elsewhere.

File: chainercv/links/model/mobilenet/tf_convolution_2d.py (hunk 12 above)
Comment (okdshin) [reply chainercv/links/model/mobilenet/tf_convolution_2d.py]: No. Chainer's convolution doesn't support asymmetrical pad. So we can not eliminate this pad.

**Comment by yuyu2172** [comment]
pfnCI, test this please

**Comment by okdshin** [comment]
@yuyu2172 Could you run tests?

**Comment by yuyu2172** [comment]
pfnCI, test this please

**Comment by yuyu2172** [comment]
Thank you. LGTM
